ambry-store/src/main/java/com.github.ambry.store/DiskSpaceAllocator.java

```diff
@@ -0,0 +1,335 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+class DiskSpaceAllocator {
+  private static final Logger LOGGER = LoggerFactory.getLogger(DiskSpaceAllocator.class);
+  private static final String RESERVE_FILE_PREFIX = "reserve_";
+  private static final FileFilter RESERVE_FILE_FILTER =
+      pathname -> pathname.isFile() && pathname.getName().startsWith(RESERVE_FILE_PREFIX);
+  private static final FileFilter FILE_SIZE_DIR_FILTER = pathname -> {
+    try {
+      return pathname.isDirectory() && Long.parseLong(pathname.getName()) >= 0;
+    } catch (NumberFormatException e) {
+      return false;
+    }
+  };
+
+  private final ReserveFileMap reserveFiles;
+  private final File reserveDir;
+  private final long requiredSwapSegmentsPerSize;
+  private volatile boolean poolInitialized = false;
+
+  /**
+   * @param reserveDir the directory where reserve files will reside. If this directory does not exist yet, it will
+   *                   be created.
+   * @throws StoreException
+   */
+  DiskSpaceAllocator(File reserveDir, long requiredSwapSegmentsPerSize) throws StoreException {
+    this.reserveDir = reserveDir;
+    this.requiredSwapSegmentsPerSize = requiredSwapSegmentsPerSize;
+    try {
+      prepareDirectory(reserveDir);
+      reserveFiles = inventoryExistingReserveFiles();
+    } catch (Exception e) {
+      throw new StoreException("Could not load from reserve directory.", e, StoreErrorCodes.Initialization_Error);
+    }
+  }
+
+  /**
+   * @param requirementsList a list of {@link DiskSpaceRequirements}s objects that describe the number and segment size
+   *                         needed by each store.
+   * @throws StoreException
+   */
+  void initializePool(Collection<DiskSpaceRequirements> requirementsList) throws StoreException {
+    reserveFiles.rwLock.writeLock().lock();
+    try {
+      try {
+        Map<Long, Long> overallRequirements = getOverallRequirements(requirementsList);
+        deleteUnneededSegments(overallRequirements);
+        addRequiredSegments(overallRequirements);
+        // TODO fill the disk with additional swap segments
+        poolInitialized = true;
+      } catch (Exception e) {
+        throw new StoreException("Exception while initializing DiskSpaceAllocator pool", e,
+            StoreErrorCodes.Initialization_Error);
+      }
+    } finally {
+      reserveFiles.rwLock.writeLock().unlock();
+    }
+  }
+
+  /**
+   * Allocate a file, that is, take a file matching the requested size from the reserve directory and move it to the
+   * provided file path
+   * @param destinationFile the file path to move the allocated file to.
+   * @param sizeInBytes the size in bytes of the requested file.
+   * @throws IOException
+   */
+  void allocate(File destinationFile, long sizeInBytes) throws IOException {
+    if (!poolInitialized) {
+      LOGGER.debug("Allocating segment before pool is fully initialized to " + destinationFile.getAbsolutePath());
```

Comment:
Changed to info.
@nsivabalan this is not an error. It will happen normally when a Log is allocating its first segment, which happens before `initializePool` is called.